src/java/org/apache/cassandra/db/SSTableImporter.java

```diff
@@ -97,7 +101,39 @@ synchronized List<String> importNewSSTables(Options options)
                         try
                         {
                             abortIfDraining();
-                            verifySSTableForImport(descriptor, entry.getValue(), options.verifyTokens, options.verifySSTables, options.extendedVerify);
+
+                            if (options.failOnMissingIndex)
+                            {
+                                Index.Group saiIndexGroup = cfs.indexManager.getIndexGroup(StorageAttachedIndexGroup.GROUP_KEY);
+                                if (saiIndexGroup != null)
+                                {
+                                    IndexDescriptor indexDescriptor = IndexDescriptor.create(descriptor,
+                                                                                             cfs.getPartitioner(),
+                                                                                             cfs.metadata().comparator);
+
+                                    String keyspace = cfs.getKeyspaceName();
+                                    String table = cfs.getTableName();
+
+                                    if (!indexDescriptor.isPerSSTableIndexBuildComplete())
+                                        throw new IllegalStateException(String.format("Missing SAI index to import for SSTable %s on %s.%s",
+                                                                                      indexDescriptor.sstableDescriptor.toString(),
+                                                                                      keyspace,
+                                                                                      table));
+
+                                    for (Index index : saiIndexGroup.getIndexes())
+                                    {
+                                        IndexIdentifier indexIdentifier = new IndexIdentifier(keyspace, table, index.getIndexMetadata().name);
+                                        if (!indexDescriptor.isPerColumnIndexBuildComplete(indexIdentifier))
+                                            throw new IllegalStateException(String.format("Missing SAI index to import for index %s on %s.%s",
+                                                                                          index.getIndexMetadata().name,
+                                                                                          keyspace,
+                                                                                          table));
```

Comment:
nit: You could use the `logMessage()` helpers from `IndexDescriptor` and `IndexIdentifier` for the exceptions above, but not critical. It would give you the standardized prefix (which we did to make it easier to search for all SAI-related logging) but that's it.

Reply:
I dont like to mix sai specific logging stuff with sstable importer. Seems totally unrelated really ...